### PR TITLE
client: return EOPNOTSUPP for fallocate with mode 0

### PR DESCRIPTION
**Pull request by mchangir**
Fixes: https://tracker.ceph.com/issues/68026
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by mchangir** [comment]
for doc updates, please see the `cephfs.pyx` file change

**Comment by gregsfortytwo** [comment]
This should get a PendingReleaseNote since we're changing the API

**Comment by mchangir** [comment]
* rebased
* added PendingReleaseNote

**Comment by mchangir** [comment]
@anthonyeleven please review `PendingReleaseNotes`

**Comment by vshankar** [comment]
oh, I tried this on a vstart cluster with strace'ing fallocate on ceph-fuse:

```
openat(AT_FDCWD, "./test", O_RDWR|O_CREAT, 0666) = 3
fallocate(3, 0, 0, 10485760)            = 0
fsync(3)                                = 0
```

`mode` is 0 and only a single rados object corresponding to the file is allocated on the data pool with `size:0`

```
./bin/rados -p cephfs.a.data stat 10000000000.00000000
cephfs.a.data/10000000000.00000000 mtime 2024-09-19T09:50:28.000000+0000, size 0
```

This is handle in kclient as per: https://github.com/ceph/ceph-client/commit/bddff633ab7bc60a18a86ac8b322695b6f8594d0 (taken from the mail you sent to ceph-users, subject "RFC: cephfs fallocate" - mentioning it here for brevity).

File: PendingReleaseNotes
Comment (vshankar) [inline PendingReleaseNotes:16]: nit: just mention that the default case is mode:0

Comment (mchangir) [reply PendingReleaseNotes]: updated the `PendingReleaseNotes` with `mode == 0`

**Comment by anthonyeleven** [comment]
Docs lgtm, note to self to wordsmith after merge.

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68179](https://tracker.ceph.com/issues/68179).

**Comment by mchangir** [comment]
* rebased
* added workunit for testing fallocate mode 0

**Comment by mchangir** [comment]
* rebased

**Review by vshankar — CHANGES_REQUESTED** [review]
The python binding tests needs fixing. See:

[vshankar-2024-09-24_04:11:42-fs-wip-vshankar-testing-20240923.114534-debug-testing-default-smithi](https://pulpito.ceph.com/vshankar-2024-09-24_04:11:42-fs-wip-vshankar-testing-20240923.114534-debug-testing-default-smithi/)/7917550

```
2024-09-24T08:52:17.447 INFO:tasks.workunit.client.0.smithi059.stdout:
2024-09-24T08:52:17.447 INFO:tasks.workunit.client.0.smithi059.stdout:=================================== FAILURES ===================================
2024-09-24T08:52:17.447 INFO:tasks.workunit.client.0.smithi059.stdout:________________________________ test_fallocate ________________________________
2024-09-24T08:52:17.448 INFO:tasks.workunit.client.0.smithi059.stdout:
2024-09-24T08:52:17.448 INFO:tasks.workunit.client.0.smithi059.stdout:testdir = None
2024-09-24T08:52:17.448 INFO:tasks.workunit.client.0.smithi059.stdout:
2024-09-24T08:52:17.448 INFO:tasks.workunit.client.0.smithi059.stdout:    def test_fallocate(testdir):
2024-09-24T08:52:17.448 INFO:tasks.workunit.client.0.smithi059.stdout:        fd = cephfs.open(b'/file-fallocate', 'w', 0o755)
2024-09-24T08:52:17.448 INFO:tasks.workunit.client.0.smithi059.stdout:        assert_raises(TypeError, cephfs.fallocate, b'/file-fallocate', 0, 10)
2024-09-24T08:52:17.449 INFO:tasks.workunit.client.0.smithi059.stdout:>       cephfs.fallocate(fd, 0, 10)
2024-09-24T08:52:17.449 INFO:tasks.workunit.client.0.smithi059.stdout:
2024-09-24T08:52:17.449 INFO:tasks.workunit.client.0.smithi059.stdout:../../../clone.client.0/src/test/pybind/test_cephfs.py:613:
2024-09-24T08:52:17.449 INFO:tasks.workunit.client.0.smithi059.stdout:_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2024-09-24T08:52:17.449 INFO:tasks.workunit.client.0.smithi059.stdout:
2024-09-24T08:52:17.449 INFO:tasks.workunit.client.0.smithi059.stdout:>   ???
2024-09-24T08:52:17.450 INFO:tasks.workunit.client.0.smithi059.stdout:E   cephfs.OperationNotSupported: fallocate failed: Operation not supported [Errno 95]
2024-09-24T08:52:17.450 INFO:tasks.workunit.client.0.smithi059.stdout:
2024-09-24T08:52:17.450 INFO:tasks.workunit.client.0.smithi059.stdout:cephfs.pyx:955: OperationNotSupported
```

**Comment by mchangir** [comment]
* rebased
* update to pybind test for mode 0

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68382](https://tracker.ceph.com/issues/68382).

File: src/test/pybind/test_cephfs.py
Comment (vshankar) [inline src/test/pybind/test_cephfs.py:613]: This should be `libcephfs.OperationNotSupported`

**Review by vshankar — CHANGES_REQUESTED** [review]
@mchangir PTAL

File: src/test/pybind/test_cephfs.py
Comment (mchangir) [reply src/test/pybind/test_cephfs.py]: fixed the exception name

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68578](https://tracker.ceph.com/issues/68578).

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20241016135728-debug

**Comment by vshankar** [comment]
@mchangir could you rebase this please? (minor pending release notes conflict).

**Comment by mchangir** [comment]
rebased

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test windows

File: PendingReleaseNotes
Comment (phlogistonjohn) [inline PendingReleaseNotes:47]: @mchangir - Sorry for the late remark but I just found out about this change. This change is causing a failure in the go-ceph ci (see https://github.com/ceph/go-ceph/issues/1044 ). The change itself is not a problem, we will deal with it. My request is to tweak the release note because 'fuse client' is a bit misleading as this change doesn't only impact fuse. Maybe tweak it to say 'fuse client or other libcephfs clients'?  Thanks.

**Pull request merged**
